### PR TITLE
feat(encoder): provide TryFrom<parser types> for encoder types

### PR DESCRIPTION
**Pull request by goto-bus-stop**
Provides `TryFrom` impls for the apollo-encoder AST types.

The conversion effectively assumes that the apollo-parser AST is valid and complete. Otherwise you get an Err, but not a very useful one, because the TryFrom impl doesn't know anything about its parent (so we can't show source code where the error originated for example).

I put it behind a feature flag so apollo-encoder doesn't *have* to depend on apollo-parser, not sure if that is truly worth it, so i also enabled it by default 😛 .

### CODE REVIEW

File: crates/apollo-encoder/Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,14 @@ categories = [
 ]
 edition = "2021"
 
+[dependencies]
+apollo-parser = { version = "0.2.12", optional = true }
+thiserror = "1.0.37"
+
+[features]
+default = ["from-parser"]
+from-parser = ["apollo-parser"]
```
Comment (lrlna) [inline crates/apollo-encoder/Cargo.toml:28]: should we just call this feature `apollo-parser`?

**Review by lrlna — APPROVED** [review]
yay! excited for this feature!!

File: crates/apollo-encoder/Cargo.toml (hunk 1 above)
Comment (goto-bus-stop) [reply crates/apollo-encoder/Cargo.toml]: i'm not sure what the general practice is on this in rustland 🤔 I used `from-parser` to describe what the feature actually does. but, if there are other features in the future that also rely on apollo-parser, it's fine if the TryFrom impls are made available anyways, so maybe that is actually counterproductive.

Comment (lrlna) [reply crates/apollo-encoder/Cargo.toml]: I am leaning towards having this TryFrom be always available. It's part of the same ecosystem, so it makes sense.

Comment (goto-bus-stop) [reply crates/apollo-encoder/Cargo.toml]: ok, I think then the current way is fine, as it's enabled by default but you can opt-out by setting `default-features=false` if you really only need the encoder and don't want the extra dependencies.

**Comment by goto-bus-stop** [comment]
One thing I noticed while working on this, is that the `Display` impls on AST nodes make it easy to do the wrong thing. I was doing `.name().to_string()` all over the place but that actually returns the source text including possible whitespace, not only the name. The same goes for StringValue nodes, where `node.to_string()` includes quotes. I think accessing the source code is a bit of a specialised operation that doesn't need to be so easy. I think we should either remove the Display impl for parser nodes entirely or replace it by a named method like `.source_string()`.

File: crates/apollo-encoder/Cargo.toml (hunk 1 above)
Comment (goto-bus-stop) [reply crates/apollo-encoder/Cargo.toml]: decided to do tht^ in slack : )

**Pull request merged**